### DISCUSSION TITLE
Middleware next?

### DISCUSSION THREAD

**Discussion by delaneyj** [trpc/trpc#586] [Q&A] [answered]
Unless I'm missing something you can't have a middleware do pre/post style flow ala koa... 

```ts
.middleware(async ({ ctx, next }) => {
  const auth = checkForAuth(ctx)
  if (!auth) return handleUnauth(ctx)
  await next()
  const update = checkForUpdatedAuth(ctx)
  if (update) {
    ctx.res.headers['x-access-token'] = update.token
  }
})
```

**Comment by KATT** [comment]
You are correct. 

I actually considered calling middlewares something else when I made them as it is not the same as express middlewares. Originally it was called liked `preHook` or something like that, but ended up being called `middleware. See the original PR https://github.com/trpc/trpc/pull/87

- What would you want to use `next()`-calling for? Is there a need for a `connect`-like middleware layer in tRPC?
- I agree `middleware` isn't the perfect word for this implementation - I'd be up for changing it - suggestions?

**Comment by delaneyj** [comment]
Well in my exact situation I'm parsing a paseto token in the preNext, and in the postNext looking at the expiration date and if its within a bound using the refreshToken to regen the accessToken and send it back in headers.  While you could do this outside of tRPC, you can't if you want to be protocol agnostic.  You could do a pre/post hook, but as Koa has shown, just having async/await cleans thing up and keep stuff in logical places.

**Comment by KATT** [comment]
I'd do both of those things in `createContext` (and attach the user obj to ctx so it's passed through to all resolvers).

Doesn't seem like you _need_ do that token regen after resolving the request, or am I misunderstanding?

Secondly, createContext is called only once per request even if you're batching, but mutating as part of a middleware will be multiplied with the number of calls you did in that req and might produce some odd side effects.

**Comment by delaneyj** [comment]
right now your encode/decode goes through transformer.   but it could be a generic middleware if it was more like koa.  also supporting things like brotli over the wire would just be another middleware.

**Comment by KATT** [comment]
The goal wasn't really for tRPC to be completely standalone, but something you can plug in to whatever server you used. I first created the createHttpServer for the integration testing, exposing it was just a _"meh, why not?"_ side-effect. 

Anyway, middlewares on a router level will not be the solution for your auth as they can be called several times in the same request (batching). Do this stuff in a connect middleware outside of tRPC or in your `createContext`.

I won't change the current implementation - but I understand the naming is confusing, so open for suggestions on that.

**Comment by mmkal** [comment, marked as answer]
@KATT I have a couple more use cases for koa-style middleware, they simplest example being query time logging:

```ts
.middleware(async ({ ctx, path, next }) => {
  const start = Date.now()
  await next()
  const end = Date.now()
  console.log({path, ctx, took: end - start})
})
```

We also have some utils using `async_hooks` which would let us put things in AsyncLocalStorage and get them in the resolver this way. I don't think this is possible with the beforewares, but curious if you have any suggestions. 

There are other things that can be done like wrapping `await next()` in a try/catch and doing error handling without needing special handlers.

**Reply by KATT** [reply]
I agree it would be useful. 

Happy to merge & make this part of a major version change if someone wanted to do the work, it's non-trivial to do

**Reply by KATT** [reply]
Turned this back into an issue at #717

**Comment by octet-stream** [comment]
@KATT I also have a scenario where doing something past `next()` call might be useful: In my case, I need to replace procedure result with cached data, if exists. I'm working on template project for Next.js 13 `/app` directory. With these new APIs they decided that you may need to perform several queries within your route segment (e. g. in `page` or `layout`), and within one file (e. g. `generateMetadata` and server component that render your page). As far as I know thay do cache and deduplicate requests made with `fetch` function, but the solution I ended up with is use `router.createCaller` function, so that I don't need to use network, because both API and application is within one Next.js project. Which means in terms of caching and queries deduplication I'm on my own, so I need to use middleware result to cache it, and when route is queried again, the data should be returned from cache. For that I made a middleware, like this one:

```ts
export const withCache = t.middleware(async ({ctx, next, path, rawInput, meta, type}) => {
  // Cache only query procedures called with `createCaller` helper
  const enableCache = type === "query" && isCreateCallerContext(ctx)

  if (enableCache) {
    const cached = cache.get({path, rawInput, meta})
    if (cached) {
      // Return cached data, if exists
      // tRPC expects middleware to return an object with following properties
      // Maybe this can be done using some sort of `overrideResult` helper function, which tRPC can take care of?
      return {
        ok: true, data: cached, marker: "middlewareMarker"
      } as any
    }
  }

  const result = await next({ctx: {...ctx, cache}})

  if (result.ok === true && enableCache) {
    cache.set({path, rawInput, meta}, (result as any).data)
  }

  return result
})
```

Then, in `page.tsx` I call a procedure:

```tsx
// Notice that `getNote` is called twice in this file, so it fires 2 queries. That's where I need cache.
const getNote = createCaller((trpc, id: string) => trpc.note.getById({id}))

export async function generateMetadata({params}: Props): Promise<Metadata> {
  const {title} = await getNote(params.id)

  return {
    title
  }
}

const NoteViewPage: AFC<Props> = async ({params}) => {
  const note = await getNote(params.id)

  // ... component implementation ...
}
```

[`type-graphql`](https://typegraphql.com/docs/middlewares.html#intercepting-the-execution-result) has similar behaviour for their middlewares. I think this might be useful too.

----

Also, I noticed that it does seem that output validation happens right after procedure returns its result. Am I right? Because I get an error when I return cached value, telling me that react can't serialize some of the result properties, if I don't manually serialize the result using same `zod` type as the procedure's result.

**Reply by KATT** [reply]
(love your username @octet-stream 🙂)

All middlewares run in order, so it won't run if you return early like you're doing here.

I feel like the code here would work and I'd be open to make it so you don't need to do `as any`.

Can you open an issue / feature req with some sort of reproduction in a stackblitz or something? I'd like to play with this and understand your usecase better

**Reply by octet-stream** [reply]
Would [my template](https://github.com/octet-stream/next-app-mikro-orm-trpc-template/tree/feature/procedure-cache) be a good for repro? It already has everything set up. I can explain in the issue how start it and where to look.
